cli/js/lib.deno.ns.d.ts

```diff
@@ -1505,47 +1509,48 @@ declare namespace Deno {
    * `Uint8Array`.
    *
    * Resolves to the number of bytes written and the remote address. */
-  export function recvfrom(rid: number, p: Uint8Array): Promise<[number, Addr]>;
+  //   export function recvfrom(rid: number, p: Uint8Array): Promise<[number, Addr]>;
 
   /** **UNSTABLE**: new API, yet to be vetted.
    *
    * A generic transport listener for message-oriented protocols. */
-  export interface UDPConn extends AsyncIterator<[Uint8Array, Addr]> {
+  export interface DatagramConn<T> extends AsyncIterator<[Uint8Array, T]> {
     /** **UNSTABLE**: new API, yet to be vetted.
      *
      * Waits for and resolves to the next message to the `UDPConn`. */
-    receive(p?: Uint8Array): Promise<[Uint8Array, Addr]>;
+    receive(p?: Uint8Array): Promise<[Uint8Array, T]>;
     /** UNSTABLE: new API, yet to be vetted.
      *
      * Sends a message to the target. */
-    send(p: Uint8Array, addr: UDPAddr): Promise<void>;
+    send(p: Uint8Array, addr: T): Promise<void>;
     /** UNSTABLE: new API, yet to be vetted.
      *
      * Close closes the socket. Any pending message promises will be rejected
      * with errors. */
     close(): void;
     /** Return the address of the `UDPConn`. */
-    readonly addr: Addr;
-    [Symbol.asyncIterator](): AsyncIterator<[Uint8Array, Addr]>;
+    readonly addr: T;
+    [Symbol.asyncIterator](): AsyncIterator<[Uint8Array, T]>;
   }
 
   /** A generic network listener for stream-oriented protocols. */
-  export interface Listener extends AsyncIterator<Conn> {
+  export interface Listener<T> extends AsyncIterator<Conn<T>> {
     /** Waits for and resolves to the next connection to the `Listener`. */
-    accept(): Promise<Conn>;
+    accept(): Promise<Conn<T>>;
     /** Close closes the listener. Any pending accept promises will be rejected
      * with errors. */
     close(): void;
     /** Return the address of the `Listener`. */
-    readonly addr: Addr;
-    [Symbol.asyncIterator](): AsyncIterator<Conn>;
+    readonly addr: T;
+
+    [Symbol.asyncIterator](): AsyncIterator<Conn<T>>;
   }
 
-  export interface Conn extends Reader, Writer, Closer {
+  export interface Conn<T> extends Reader, Writer, Closer {
     /** The local address of the connection. */
-    readonly localAddr: Addr;
+    readonly localAddr: T;
     /** The remote address of the connection. */
-    readonly remoteAddr: Addr;
+    readonly remoteAddr: T;
     /** The resource ID of the connection. */
```

Comment:
I'm a bit weary of making Conn generic. Maybe it's okay but  I wonder if it could be avoided...

Reply:
A solution could be to replace T with Addr, where Addr is:
```
interface Addr {
  transport: Transport;
  hostname?: string;
  port?: number;
  address?: string;
}
```
I had it like this on an early stage, but I found it less strict than having a generic type. This way the `addr` of each Conn depends on the arguments used on `listen`.

(cc @kitsonk - for expert TS advice)

Reply:
This is probably why Go uses a string for hostname and port. It's nice to just have a single string for the address...

Reply:
Well Go has a single Addr type (https://golang.org/pkg/net/#Addr), this way I could replace all Transport specific Addr to:
```
interface Addr {
    transport: string;
    address: string;
}
```
But we would no longer be able to access `hostname` nor `port` without doing `address.split(":")`. Is this acceptable?

Reply:
A compromise @bartek, @ry and yours truly came up with on our call:

```ts
export interface Conn extends Reader, Writer, Closer {
    /** The local address of the connection. */
    readonly localAddr: TCPAddr | UnixAddr;
```

Reply:
> 
> 
> A compromise @bartek, @ry and yours truly came up with on our call:
> 
> ```ts
> export interface Conn extends Reader, Writer, Closer {
>     /** The local address of the connection. */
>     readonly localAddr: TCPAddr | UnixAddr;
> ```

Alright! Pushing changes soon!

Reply:
By the way, why Addr needs a transport field? An address is an address, isn't it?

What if I wanted to use the same address between UDP and TCP and maybe QUIC?